[#1] 회원가입

qxent300 opened the pull request:
## 회원가입 기능 구현
- 데이터를 입력받아 DB 테이블에 저장

qxent300 committed:
회원가입 기능 구현

f-lab-neo commented on .gitattributes:
```diff
@@ -0,0 +1,3 @@
+/gradlew text eol=lf
+*.bat text eol=crlf
+*.jar binary
```
프로젝트 init과 관련된 코드들은, 기능 pr과 분리해서 올려주시는게 좋습니다!
보통은 init의 경우는 pr로 관리하지 않고 처음 repository를 생성할 때 함께 main 브랜치에 넣어두는 편 입니다!

qxent300 replied:
네 알겠습니다.

f-lab-neo commented on .gitignore:
```diff
@@ -0,0 +1,37 @@
+HELP.md
+.gradle
+build/
+!gradle/wrapper/gradle-wrapper.jar
+!**/src/main/**/build/
+!**/src/test/**/build/
+
+### STS ###
+.apt_generated
+.classpath
+.factorypath
+.project
+.settings
+.springBeans
+.sts4-cache
+bin/
+!**/src/main/**/bin/
+!**/src/test/**/bin/
+
+### IntelliJ IDEA ###
+.idea
+*.iws
+*.iml
+*.ipr
+out/
+!**/src/main/**/out/
+!**/src/test/**/out/
+
+### NetBeans ###
+/nbproject/private/
+/nbbuild/
+/dist/
+/nbdist/
+/.nb-gradle/
```
.gitignore 파일이 무슨 역할을 할까요!?

qxent300 replied:
git 버전 관리시 제외할 파일 목록을 관리합니다.
한가지 궁금한 내용이 있습니다.
실제로 이 파일을 사용하는 경우가 존재한가요?
버전 관리를 하지 않겠다는 뜻이 결국 불필요한 파일이라는 뜻인데, 개발자가 프로젝트를 완성하는 과정에서 불필요한 파일을 생성하는 경우가 있는걸까요?
만약 있더라도 해당 파일을 삭제하면 되지 않을까요?

f-lab-neo replied:
실제로 사용됩니다.
위  파일들이 제외되는 이유는, 실제 프로젝트에 필요한 내용이라기 보다는 ide나 git과 같은 프로젝트를 관리를 위한 데이터이기 떄문입니다. (보통 개인적인 설정들을 저장하는 파일)

예를들어 intellij에서 동식님이 개인적으로 설정을 바꿨습니다. (이런 정보는  .intellij 폴더에 저장됩니다)
그리고 그게 commit되어 repository에 저장됩니다.

그리고 제가 그 저장소에서 pull받으면 동식님의 개인적인 설정을 제가 따라가게 됩니다. (제 설정을 잃어버립니다.)
그리고 제가 또 개인적인 설정을 바꾸고 올리게 되면... 이런 상황이 반복되고, 코드리뷰시에도 불필요한 파일이 계속 change되어 올라오게 됩니다.

이런 상황을 막기 위해, 정말 프로젝트를 구성하기 위한 파일만 commit되도록 위와같은 설정을 하는 것 입니다.

qxent300 replied:
그러면 개인적인 설정 파일같은 경우에는 commit 목록에서 제외하면 되는데 혹시 모를 상황에 대비하여 .gitignore 에 등록해둔다 라고 이해하면 될까요?

f-lab-neo replied:
넵 맞습니다~
실제로 작업을 하다보면 많은 양의 파일을 수정하고 커밋하고 될텐데, 커밋에 포함하지 않을 파일을 하나하나 신경쓰는 수고를 덜기 위함도 있습니다!

qxent300 replied:
네 알겠습니다.
감사합니다.

f-lab-neo commented on build.gradle:
```diff
@@ -0,0 +1,33 @@
+
+dependencies {
+	implementation 'org.springframework.boot:spring-boot-starter-web'
+    implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
+    implementation 'org.springframework.boot:spring-boot-starter-jdbc'
+    implementation 'com.mysql:mysql-connector-j'
+    runtimeOnly 'com.mysql:mysql-connector-j'
+	testImplementation 'org.springframework.boot:spring-boot-starter-test'
+	testRuntimeOnly 'org.junit.platform:junit-platform-launcher'
```
dependencies에는
implementation, api, runtimeOnly 등등.. 여러가지가 있는데요, 이게 무엇을 뜻하는지 알아보면 이 설정이 무엇을 의미하는지 이해할 수 있을거에요~

qxent300 replied:
의미를 찾아보았는데

implenetation : 컴파일과 런타임(실행)에 모두 필요
runtimeOnly : 런타임(실행)에만 필요

라고 나와있는데, 그러면 위의 build.gradle 파일에 com.mysql:mysql-connector-j 부분이 implementation과 runtimeOnly에 중복으로 되어있는 부분을 runtimeOnly는 삭제해도 되는걸까요?

f-lab-neo replied:
넵 맞습니다~ 보통 connector는 runtimeOnly로 설정하는데 implementation으로 설정된게 이상하긴 하네요~

implenetation, runtimeOnly 외에도 여러가지 더 많이 있는데 의미를 찾아보면 좋을 것 같아요.

추가적으로, 런타임, 컴파일타임 이라는것은 무엇일까요?

qxent300 replied:
소스 코드를 기계어로 변환하는 과정이 컴파일 타임이고, 컴파일된 결과를 실행하면 런타임이 됩니다.

f-lab-neo commented on build.gradle:
```diff
@@ -0,0 +1,33 @@
+	}
+}
+
+repositories {
+	mavenCentral()
+}
```
repositories는 무엇이면 maven이 무엇일까요?

qxent300 replied:
mavenCentral() 에 대해 찾아보았는데 Maven Central Repository에 접속하여 필요한 의존성을 가져오는 함수 라고 나와있습니다.
관련하여 질문이 있습니다.
mavenCentral() 은 메서드 형태인데, 메서드라면 소스로 구현이 되어있다는 뜻인데 설명에 나와있는 '접속한다' 는 표현이 어떻게 나온건지 잘 모르겠습니다.

f-lab-neo replied:
build.gradle의 dependencies에 implementation 등으로 라이브러리 이름과 버전을 명시하면, gradle은 어딘가에서 그 라이브러리를 찾아서 다운받고 설치해야합니다.
mavenCentral은 그런 저장소의 가장 기본 (https://repo.maven.apache.org/maven2) 에서 그 라이브러리를 찾으라고 명시하는 역할을 합니다.
즉, 실제로 접속하여 필요한 의존성을 가져온다기 보다는, 접속할 곳을 알려주는 함수라고 생각해 주시면 됩니다.

말씀 주신 것 처럼, mavenCenteral은 결국 코드기때문에 포커싱 하고 ctrl + b 누르면 따라들어갈 수 있습니다. 직접 코드를 보시면 어떤 이야기인지 알 수 있을거에요.

maven의 repository란 뭔가요?

qxent300 replied:
maven 이라는 자바 프로젝트 관리 도구에서 사용하는 라이브러리 등을 관리하는 시스템이 repository 입니다.
한가지 질문이 있습니다.
gradle의 특징 중에 '설정 주입 방식을 사용하므로 재사용에 용이하다' 는 글을 보았는데, 이 뜻을 잘 모르겠습니다.
설정을 재사용 한다는 말의 의미를 잘 모르겠습니다.
<img width="810" height="327" alt="image" src="https://github.com/user-attachments/assets/3e92d426-112a-463d-a579-1e5322a3be32" />

f-lab-neo replied:
지금은 사용하지 않을것이지만, 나중에 멀티모듈 프로젝트를 하다보면 (gradle 모듈의 개념을 아셔야 하는데 지금은 모르셔도 괜찮습니다) 모듈마다 공통 설정을 해줘야 하는 경우가 있습니다.
이때 최상위 gradle에 정의된 설정을 재활용하여 공통설정으로 넣을수가 있어요. 이런 경우를 말하는 겁니다!

qxent300 replied:
네 알겠습니다.
멀티 모듈 프로젝트에 대해서는 제가 따로 찾아보겠습니다.
감사합니다.

f-lab-neo commented on build.gradle:
```diff
@@ -0,0 +1,33 @@
+plugins {
+	id 'java'
+	id 'org.springframework.boot' version '3.5.5'
```
스프링 부트 3.5.5는 무엇이 새로 생겼는지, release note를 한번 찾아볼까요!?

qxent300 replied:
https://github.com/spring-projects/spring-boot/releases/tag/v3.5.5 입니다.
한가지 질문이 있습니다.
해당 수정 내용에 대한 담당자가 전부 다른 사람들이 나오는데, 오픈소스라서 아무나 수정할 수 있는 걸까요?

f-lab-neo replied:
넵 아무나 수정할 수 있습니다.
다만 오픈소스를 관리하는 사람들이 있고. 관리자가 아닌 사람들이 github issue에 문제를 작성하거나, 다른사람이 작성한 issue를 보고 pr을 올리면 관리자가 해당 코드를 검수하고 코드에 반영합니다.

저희가 진행하고있는 코드리뷰와 동일한 프로세스로 오픈소스가 만들어지고 있고 서비스회사를 가서도 마찬가지입니다!

qxent300 replied:
네 알겠습니다.
감사합니다.

f-lab-neo commented on src/main/java/com/project/acre/controller/MemberController.java:
```diff
@@ -0,0 +1,43 @@
+import org.springframework.web.bind.annotation.PostMapping;
+
+@Controller
+public class MemberController {
```
controller를 어떻게 정의할 수 있을까요?
MVC 에서 controller가 무엇일까요!?

qxent300 replied:
MVC 에서의 Controller는 Model과 View 사이의 흐름을 제어하는 역할입니다.

f-lab-neo commented on src/main/java/com/project/acre/controller/MemberController.java:
```diff
@@ -0,0 +1,43 @@
+public class MemberController {
+    private final MemberService memberService;
+
+    @Autowired
```
Autowired가 무엇을 해주는 어노테이션인가요?
여기에 이 어노테이션을 붙인 이유는 무엇인가요?

qxent300 replied:
의존성을 주입하는 어노테이션입니다.
위 코드는 생성자 주입 방법을 적용하였고, 해당 방법이 스프링에서 권장되는 방법이라고 하여 적용하였습니다.

f-lab-neo replied:
왜 위 방식이 가장 권장될까요?
constructor에 @Autowired가 없으면 동작 할까요?

qxent300 replied:
setter 주입이나 필드 주입과 다르게 생성자 주입 방법을 사용하면 생성자를 호출하는 시점(객체가 실제로 생성되는 시점)에 의존성이 주입되어 안전하다고 알고 있습니다.
그리고 생성자가 하나만 있는 경우에는 @Autowired를 생략하여도 스프링 컨테이너가 알아서 생성자에 의존성을 주입합니다.

f-lab-neo replied:
좋은 답변이에요!
의존성 순환 문제에 대해서도 같이 공부해보면 좋을 것 같습니다!

qxent300 replied:
네 알겠습니다.
감사합니다.

f-lab-neo commented on src/main/java/com/project/acre/controller/MemberController.java:
```diff
@@ -0,0 +1,43 @@
+        this.memberService = memberService;
+    }
+
+    @GetMapping("/members/login")
```
rest api란 무엇인가요? get, post, delete, put, fetch가 무엇을 의미하는지, rest api를 기반으로 한번 알아보아요!

qxent300 replied:
자원(데이터)의 처리를 HTTP Method를 통해  CRUD를 적용한 인터페이스를 Rest API 라고 이해하였습니다.

Create : POST
Read : GET
Update : PUT
Delete : DELETE

인것 같습니다.

fetch 는 비동기방식을 처리하는 자바스크립트 메소드 라고 나와있습니다.

한가지 질문이 있습니다.
제가 알기로는 url에 파라미터 값을 담아 전달하는 GET이 아니면 나머지는 전부 POST 방식으로 알고 있는데, PUT과 DELETE를 사용해야하는 이유가 따로 있을까요?
결국 데이터 처리는 서버로 받아와서 Service 객체에서 진행할텐데, POST로 그냥 사용해도 되지 않을까요?

f-lab-neo replied:
fetch가 아니라 patch인데 제가 오타를 썻군요 ㅎㅎ;; 한번 알아봐 주시면 좋습니다!

>제가 알기로는 url에 파라미터 값을 담아 전달하는 GET이 아니면 나머지는 전부 POST 방식으로 알고 있는데, PUT과 DELETE를 사용해야하는 이유가 따로 있을까요?

말씀하신대로 POST로 모두 처리해도 문제는 없습니다.
실제로도 실무적으로 데이터 요청 api의 파라메터가 너무 많거나 복잡하여 GET을 사용할 수 없다면 POST를 사용하기도 합니다.

rest api를 공부해보시면, uri는 자원을 나타내고, Method는 자원을 처리하는 방법을 나타낸다고 말하고 있습니다.

즉, POST,GET, PUT ,DELETE를 사용하는것은, 실무적으로 동작 하느냐 마느냐를 따지는것이 아니라, 의미적으로 이 자원을 "어떻게 처리할 것인지" 를 표현하는 것 입니다.

위와같은 이런 약속을 일하는 사람들끼리 같은 컨텍스트로 이해하고, DELETE면, "아 지우는 api겠구나" 빠르게 의사소통 할 수 있습니다.

정리해보면.. 모두가 위와같은 이해를 통해 api를 만들고 있으니, 우리도 그 규칙을 이해하고 지키자... 정도로 이해해주시면 됩니다.

qxent300 replied:
네 알겠습니다.
프로젝트 진행시 말씀해주신 내용에 대해 생각하며 진행하겠습니다.
감사합니다.

f-lab-neo commented on src/main/java/com/project/acre/controller/MemberController.java:
```diff
@@ -0,0 +1,43 @@
+
+    @GetMapping("/members/join")
+    public String joinMember() {
+        return "members/joinMember";
```
spring 자체에서 html을 랜더링하여 내려주는 방식도, 간단한 웹 어플리케이션에서는 많이 사용되지만.
모던한 환경에서는 api는 json을 응답하고 클라이언트에서는 api call을 통해 데이터를 가져오고, 그걸 통해 동적으로 화면을 그려내는 방식을 사용합니다.

즉, 화면(client)와 서버가 따로 관리되는 형태입니다.
thymeleaf를 사용하지 않고 (thymeleaf는 무엇일까요?) 위 기능을 어떠게 구현할 수 있을까요!?

ResponseEntity에 대하여도 한번 알아보고, 위 기능을 어떻게 구현할지 고민해보세요! (3시간 이상 고민하고, 알아본 뒤 모르겠으면 저에게 문의주세요!)

qxent300 replied:
ResponseEntity 클래스를 사용하면 HTTP 상태코드, 헤더, 바디를 포함하여 반환할 수 있다고 나와있습니다.
세 가지 질문이 있습니다.


1. ResponseEntity를 사용하면 return 값은 자동으로 매핑된 url로 반환되는 것일까요?

2. 위 코드는 단순히 회원가입 페이지로 이동하는 메서드인데, 이러한 부분도 ResponseEntity를 적용하는게 더 나을까요?
    만약 적용한다면, 어떤식으로 해야 하는지 잘 모르겠습니다.
    제가 찾아본 예시 코드들을 보면 리턴값이 [ return new ResponseEntity<>(body, header, status); ] 와 같은 형식으로 되어있는데, 이 코드의 어느 부분에서 jsp 파일 경로를 확인할 수 있을까요?

3. 말씀해주신 api call을 통해 데이터를 가져와 동적으로 화면을 그려내는 방식에 대하여 이해가 잘 되지 않아 추가로 질문드립니다.
예를 들어, 회원정보를 조회하는 경우에는 ResponseEntity의 body에 회원정보 데이터를 담아서 return하면 말씀하신대로 클라이언트에서 해당 데이터를 가져와서 화면을 그려낸다는게 이해가 되는데, 이렇게 서버에서 무엇인가 데이터를 담아서 주는게 아닌 그 외의 경우에는 어떤식으로 사용하면 될지 잘 모르겠습니다.

f-lab-neo replied:
전체적으로 질문을 읽어보니, jsp와 같은 방식에 많이 익숙하신 것 같아요.
api에서 화면까지 담당하는것이 아닌, 데이터만 내려준다고 생각을 한번 바꿔볼까요!?
Dto를 api의 응답으로 넣으면 결과가 어떻게 나올까요?
http status가 api응답에서 어떻게 활용되는지 한번 생각해봐도 좋을 것 같습니다..!

- @ResponseBody에 대해서도 한번 알아볼까요?
- @RestController 에 대해서도 한번 알아볼까요?

f-lab-neo commented on src/main/java/com/project/acre/controller/MemberController.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @PostMapping("/members/join")
+    public String joinMember(MemberForm form) {
```
@RequestBody 어노테이션에 대해서 한번 알아볼까요?
위 어노테이션을 사용했을때와, 동식님이 작성해주신것과 같이 사용했을때의 차이점이 무엇이 있을까요!?

qxent300 replied:
@RequestBody 어노테이션은 json 또는 xml 형식으로 데이터를 받을때 사용한다고 나와있습니다.
제가 사용한 방식은 <form> 태그를 통해 데이터를 받을때 사용할 수 있다고 나와있습니다.
하지만 왜 사용해야 하는지에 대해서는 이유가 안나와 있어서 잘 모르겠습니다.
결국 데이터를 받아와 사용하는 것은 동일한데 @RequestBody를 사용하는 이유가 따로 있을까요?

f-lab-neo replied:
requestBody를 사용하지 않고, 위와같이 처리한다면 스프링은 내부적으로 modelAttribute를 이용합니다.
아래에 두 방식을 비교한 아티클을 보내드립니다!

https://tecoble.techcourse.co.kr/post/2021-05-11-requestbody-modelattribute/

f-lab-neo replied:
>결국 데이터를 받아와 사용하는 것은 동일한데 @RequestBody를 사용하는 이유가 따로 있을까요?

앞으로는 데이터를 주고받을 때 json을 한번 이용해볼까요!?

qxent300 replied:
네 알겠습니다.

@RequestBody나 @ModelAttribute는 동일한 결과를 가져오나 해당 결과를 가져오기 위한 과정에서 차이점이 있고, 제가 작성한 방식은 @ModelAttribute를 내부적으로 이용하며, 이는 query string 이나 form 형식이 아니면 사용할 수 없기 때문에 앞으로는 json 방식으로 데이터를 주고받고, 이와 관련하여 @RequestBody를 사용한다.

로 이해하면 될까요?

f-lab-neo replied:
넵 맞습니다~

f-lab-neo commented on src/main/java/com/project/acre/controller/MemberController.java:
```diff
@@ -0,0 +1,43 @@
+        member.setName(form.getName());
+        member.setNickname(form.getNickname());
+        member.setEmail(form.getEmail());
+        member.setPhone(form.getPhone());
```
객체지향의 사실과 오해를 읽으셨다면, 객체의 책임, 역할, 협력에 대해서 이해하셨을거에요.
member를 생성하는 이 코드가, 객체지향적으로 controller에 존재하는것에 대하여 어떻게 생각하실까요!?

qxent300 replied:
지난번 멘토링 시간에 설명해주신 내용을 바탕으로 해당 코드는 Service에서 처리하도록 다시 수정하겠습니다.

f-lab-neo commented on src/main/java/com/project/acre/controller/MemberDto.java:
```diff
@@ -0,0 +1,58 @@
+    private String nickname;
+    private String email;
+    private String phone;
+
+    public String getId() {
+        return id;
+    }
+
+    public void setId(String id) {
+        this.id = id;
+    }
+
+    public String getPassword() {
+        return password;
+    }
+
+    public void setPassword(String password) {
```
getter와 setter를 혹시 손으로 하나하나 작성하셨을까요?
그렇다면 intellij의 보일러플레이트 코드를 자동으로 작성해주는 여러 기능들이 있습니다. 한번 찾아보시고 활용해보세요!
https://jjdevelop.tistory.com/31

qxent300 replied:
감사합니다.
단축키를 조금 더 잘 활용하도록 하겠습니다.

f-lab-neo commented on src/main/java/com/project/acre/controller/MemberForm.java:
```diff
@@ -0,0 +1,58 @@
+package com.project.acre.controller;
+
+public class MemberForm {
```
Form이라는 네이밍을 사용하는것이 익숙하실 수 있습니다.
하지만 개발자로서 함께 일하려면 기본적으로 통용되는 개념과 네이밍을 이해하셔야 해요.

Dto란 무엇일까요? 이에 대해서 알아보고 Form 대신 이 클래스를 어떻게 바꿀 수 있을지 생각해보아요!

qxent300 replied:
해당 클래스는 MemberDto로 수정하겠습니다.

f-lab-neo commented on src/main/java/com/project/acre/controller/MemberDto.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    public void setId(String id) {
+        this.id = id;
```
this란 무엇인가요?

qxent300 replied:
수정 전 코드를 기준으로 MemberForm 클래스를 의미합니다.

f-lab-neo commented on src/main/java/com/project/acre/repository/MemberRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface MemberRepository {
+    Member join(Member member);
+    Optional<Member> findMember(String id, String password);
```
Optional이란 무엇인가요? 왜 사용해야하나요?

qxent300 replied:
강의 내용 중에서 Optional을 사용하면 값이 없는 경우에도 NullPointerException을 발생시키지 않는다고 설명하여 사용하였습니다.
위 코드에서는 회원정보를 찾을때, 등록된 회원정보가 없으면 빈 객체(빈 값)를 생성하는 것으로 이해하여 사용하였는데, 제가 제대로 이해한 것이 맞을까요?

f-lab-neo replied:
넵 제대로 이해하신것이 맞습니다!
Optional을 사용하면 마법처럼 NPE(Null pointer exception) 일 발생하지 않는걸까요? 왜 NPE가 발생하지 않나요?

qxent300 replied:
Optional을 사용하면 isPresent() 를 통해 null check를 하거나 orElse() 등으로 리턴값을 처리하여 NPE를 방지하는것 같습니다.
Optional을 사용하지 않는다면 try - catch 블록을 생성하여 NPE 처리 코드를 작성해야 하는데, Optional을 사용하면 간단하게 작성이 가능하여 Optional을 사용하는 것으로 이해하였는데, 맞을까요?

f-lab-neo replied:
넵 맞습니다-~!!

일반적으로 npe를 방어하기 위해서
if (a==null)
과같은 분기를 많이 사용합니다.

하지만 아이러니하게도 optional 제대로 쓰기위해 지켜야 할 항목이 너무 많아서, 사용하지 않는걸 권하는 의견도 많답니다 ㅎㅎ

f-lab-neo commented on src/main/java/com/project/acre/repository/MemberRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+import java.util.Optional;
+
+public interface MemberRepository {
```
Repository가 무엇인가요?? 왜 Repository라는 네이밍을 사용하셨을까요?

qxent300 replied:
솔직히 말씀드리면 인프런 강의에서 해당 네이밍을 사용하여 그대로 따라하였습니다.
Repository의 뜻이 저장소여서 DB와 관련된 기능을 처리하는 클래스의 명칭으로 사용하였다고 생각합니다.

f-lab-neo commented on src/main/java/com/project/acre/repository/MySQLMemberRepository.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.Optional;
+
+@Repository
+public class MySQLMemberRepository implements MemberRepository {
```
Repository라는 네이밍과 MySql이라는 네이밍이 어울리는지, Repository에 대해서 알아보고 생각을 공유해주세요!
보통, interface의 구현체는 Impl(implements)이라는 suffix를 붙여서 구현합니다.
-> MemberRepositoryImpl

qxent300 replied:
이 부분도 인프런 강의의 예제코드에서 사용한 방식을 그대로 따라하였습니다.
강의 내용을 다시 확인하였고, 위와 같이 네이밍을 한 이유가 스프링의 특성에 대한 설명을 더 쉽게 이해시키기 위하여 눈에 보이도록 구분하기 위함인 것 같습니다.
해당 코드는 말씀하신대로 Impl 을 붙이도록 수정하겠습니다.

f-lab-neo replied:
Repository 패턴에 대해서 혹시 공부하실수 있을까 하여 질문드렸습니다!
깊게 공부할 필요는 없지만, 모든 네이밍에 대해서 왜 이런 이름이 지어졌을지도 한번 공부해보시면 좋을 것 같아요!
추가적으로 MySQL이라는것이 이 클래스 이름에 있는게 적절할지도 한번 생각해보면 좋을 것 같습니다!

qxent300 replied:
네 알겠습니다.
Repository 패턴에 대해서 공부하겠습니다.
감사합니다.

f-lab-neo commented on src/main/java/com/project/acre/repository/MemberRepositoryImpl.java:
```diff
@@ -0,0 +1,50 @@
+
+        Map<String, Object> parameters = new HashMap<>();
+        parameters.put("id", member.getId());
+        parameters.put("password", member.getPassword());
```
비밀번호를 그대로 db에 저장하면 어떤 문제점이 있을까요? 어떻게 해소할 수 있을까요!?

qxent300 replied:
암호화 과정이 필요합니다.
관련하여 양방향 / 단방향 암호화가 있는 것으로 확인하였습니다.
양방향은 암호화 후 복호화가 가능하고, 단방향은 복호화가 불가능한 것으로 확인하였습니다.
예시로는 Spring Security를 사용함을 확인하였는데, 저도 동일하게 사용하면 될까요?

f-lab-neo replied:
Spring Security는 복잡하고 무겁습니다.
지금 해당 프레임워크를 사용하기는 너무 학습량이 많으며, 위와같이 간단한 암호화의 경우는 기본 클래스로도 처리 가능합니다.

비밀번호의 경우 어떤 암호화를 사용하실껀가요?

qxent300 replied:
자바에서 단방향 암호화는 SHA-256 방법을 많이 사용하는것 같아 해당 방법을 적용해보겠습니다.

f-lab-neo commented on src/main/java/com/project/acre/repository/MySQLMemberRepository.java:
```diff
@@ -0,0 +1,50 @@
+        parameters.put("email", member.getEmail());
+        parameters.put("phone", member.getPhone());
+
+        Number key = jdbcInsert.executeAndReturnKey(parameters);
+        member.setKey(key.longValue());
+
+        return member;
```
왜 member라는 객체를 이 메서드의 파라메터로 받았으며, member에 setKey를 하고 return 했나요?

qxent300 replied:
member를 파라메터로 받은 이유는 입력한 회원정보를 DB로 저장하기 위하여 member에 클라이언트에서 입력한 회원정보를 담아서 파라메터로 받았습니다.
setKey의 이유는 SimpleJdbcInsert에서 DB로 insert를 하기 위해서는 executeAndReturnKey를 사용하고, 해당 메서드가 key 값을 return 한다고 하여 위와 같이 코드를 작성하였습니다.

f-lab-neo replied:
설명주셔서 감사합니다!
개인적으로 Repository의의 책임은과 역할은 무엇인지에 대해 알아보면 좋을 것 같아서요..!
의미에 대해서 간단히 알아보고 다시한번 답변 주시면 좋을 것 같습니다!

qxent300 replied:
Repository 는 비즈니스 로직과 DB 사이에서 데이터 처리에 대한 부분을 담당하며,  Service로부터 데이터를 받아서 DB에 저장하거나 DB로부터 데이터를 읽어와서 Service로 보내주는 기능들을 모아둔 곳이라고 생각합니다.

f-lab-neo commented on src/main/java/com/project/acre/repository/MemberRepositoryImpl.java:
```diff
@@ -0,0 +1,50 @@
+        return member;
+    }
+
+    @Override
+    public Optional<Member> findMember(String id, String password) {
+        return null;
+    }
```
Optional이 return값임에도 null을 반환한 이유가 있을까요!?

qxent300 replied:
이 부분은 회원정보를 찾는 기능의 메서드이고, 아직 구현하지 못하여 임의로 null로 하였습니다.

f-lab-neo replied:
미구현이더라도, Optional이 반환이면 Optional.empty() 를 반환해주는게 좋습니다 :)
협업하는 입장에서 생각해보면 동료가 개발환경에서 해당 메서드를 사용했다가 NPE가 발생해버리면 원인을 찾느라 힘들수도 있거든요!

qxent300 replied:
네 알겠습니다.
이 부분은 말씀해주신대로 수정하겠습니다.
말씀해주신 내용에 대하여 앞으로는 미구현 코드의 틀만 구성할꺼면 리턴타입을 맞추거나, 아니면 아예 작성하지 않도록 주의하겠습니다.

f-lab-neo commented on src/main/java/com/project/acre/service/MemberService.java:
```diff
@@ -0,0 +1,25 @@
+    public Long join(Member member) {
+        memberRepository.join(member);
+        return member.getKey();
+    }
```
@Transactional 이란 무엇일까요?
그리고 트랜젝션이란 무엇일까요?

이런 save를 하는 코드에 트랜젝션이 없다면 어떤 문제가 발생할 수 있을까요?

qxent300 replied:
DB에서 데이터가 처리되는 단위를 트랜잭션이라고 알고 있습니다.
정상적인 경우 commit을, 오류가 발생하면 rollback을 처리하기 위해 @transactional 어노테이션을 사용하도록 코드 수정하겠습니다.

f-lab-neo commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,5 @@
+spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
+spring.datasource.url=jdbc:mysql://localhost:3306/acre?serverTimezone=Asia/Seoul
+spring.datasource.username=root
+spring.datasource.password=root
```
파일의 마지막에 한 줄 띄어주는것이 POSIX 표준입니다. (표준을 따르지 않아서 - 표시가 있습니다)
표준을 지켜서 코딩해주세요~

https://velog.io/@teddynu/%ED%8C%8C%EC%9D%BC-%EB%A7%88%EC%A7%80%EB%A7%89-%EC%A4%84%EC%9D%80-%EB%B9%88-%EC%A4%84%EC%9D%B4%EC%96%B4%EC%95%BC-%ED%95%A9%EB%8B%88%EB%8B%A4

qxent300 replied:
네 수정하겠습니다.

f-lab-neo commented on build/resources/main/templates/home.html:
```diff
@@ -0,0 +1,15 @@
+<!DOCTYPE html>
+<html lang="ko">
+<head>
+    <meta charset="UTF-8">
+    <title>Home</title>
+</head>
+<body>
+    <div>
+        <p>
+            <a href="members/login">로그인</a>
+            <a href="members/join">회원가입</a>
+        </p>
+    </div>
+</body>
+</html>
```
페이지의 가장 처음은 home.html보다는 index.html을 가장 많이 쓰며 표준입니다 :)

qxent300 replied:
네 알겠습니다.
index.html로 수정하겠습니다.

f-lab-neo commented on src/main/resources/templates/members/joinMember.html:
```diff
@@ -0,0 +1,33 @@
+            <button type="submit">등록</button>
+        </form>
+    </div>
+</body>
```
위에서 thymeleaf 를 사용하지 말라고 말씀드렸는데요.
그렇게되면 이 페이지를 어떻게 바꿔야 할까요? form이 필요할까요?

화면은 gpt와 같은 ai 툴을 마음껏 이용해도 됩니다. 한번 구성해보세요!

qxent300 replied:
멘토님, 말씀하신 방법을 잘 모르겠습니다.
제가 어떤 내용에 대해 찾아보면 될지 알려주시면 참고하여 코드 수정하겠습니다.

f-lab-neo replied:
api는 json으로 응답하고, 프론트엔드는 javascript에서 해당 fetch로 api를  호출하여 json 데이터를 받아옵니다.
그리고 html에 데이터를 잘 넣어줍니다.

`/api/user` 라는 api가 있다면 아래와 같이 코드를 작성할 수 있습니다 :) 

```
<!DOCTYPE html>
<html lang="ko">
<head>
  <meta charset="UTF-8">
  <title>API Fetch Example</title>
</head>
<body>
  <h1>사용자 정보</h1>
  <div id="user"></div>

  <script>
    // API 호출
    fetch("http://localhost:3000/api/user")
      .then(response => response.json())
      .then(data => {
        // HTML에 넣기
        const userDiv = document.getElementById("user");
        userDiv.innerHTML = `
          이름: ${data.name} <br>
          나이: ${data.age} <br>
          직업: ${data.job}
        `;
      })
      .catch(err => console.error(err));
  </script>
</body>
</html>
```

f-lab-neo commented on src/test/java/com/project/acre/AcreApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
+class AcreApplicationTests {
+
+	@Test
+	void contextLoads() {
+	}
+
+}
```
사용하지 않는 테스트는 제거해주세요!

qxent300 replied:
네 알겠습니다.
수정하겠습니다.

f-lab-neo commented on src/test/java/com/project/acre/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
```
spring boot test를 사용한 이유가 무엇인가요?

`유닛테스트` 와의 차이점이 무엇일까요?

qxent300 replied:
@SpringBootTest 어노테이션을 사용하면 실제 스프링을 실행하여 전체 통합테스트를 진행할 수 있다고 하여 작성하였습니다.
유닛테스트는 기능 단위별로 테스트를 진행합니다.
각 단위별 테스트를 진행하도록 수정하겠습니다.

qxent300 replied:
멘토님, @SpringBootTest 를 제거하면 DB로 저장하는 테스트를 진행할 수 없는데, 말씀하신 단위테스트는 어떤 식으로 진행하면 될까요?
단위테스트는 Java로만 구성하여 테스트를 진행하고, 나중에 통합테스트를 따로 생성하여 통합테스트에서 DB에 저장이 잘 되는지 확인할 수 있도록 테스트 코드를 생성하면 될까요?

f-lab-neo replied:
Mock이라는 키워드를 찾아보셔도 좋을 것 같습니다~
반드시 단위테스트만 진행하는게 아니라, 테스트의 목적에 맞게 알맞는 테스트 방식을 선택하는것 또한 중요합니다.
테스트의 종류에는 무엇이 있고, 각 테스트는 어떤 경우에ㅜ사용될까요!?

f-lab-neo commented on src/test/java/com/project/acre/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,28 @@
+
+        Long saveKey = memberService.join(member);
+
+        System.out.println(saveKey);
```
테스트는 결과를 눈으로 확인하기 위함이 아닌, 자동화된 테스트를 구축함에 그 목적이 있습니다.
print를 하면 테스트는 항상 성공하고, 실제로 로직에 문제기 생기더라도 개발자가 인지할 수 없습니다.

assertj를 통한 테스트를 다시 작성해주세요.

qxent300 replied:
네 알겠습니다.
assertj 로 수정하겠습니다.

f-lab-neo reviewed (changes requested):
수고하셨습니다 :)
코멘트 확인해주시고 어려운 부분있다면 편하게 문의주세요!

f-lab-neo commented on src/main/java/com/project/acre/repository/MemberRepositoryImpl.java:
```diff
@@ -0,0 +1,50 @@
+    @Override
+    public Member join(Member member) {
+        SimpleJdbcInsert jdbcInsert = new SimpleJdbcInsert(jdbcTemplate);
+        jdbcInsert.withTableName("member").usingGeneratedKeyColumns("rownum");
```
테이블 명세 (DDL) 이 없네요~ 공유 부탁 드립니다!

qxent300 replied:
네 알겠습니다.
sql 폴더 하위의 ddl.sql 파일에 작성하여 커밋하겠습니다.

f-lab-neo commented on src/main/java/com/project/acre/controller/HomeController.java:
```diff
@@ -7,6 +7,6 @@
 public class HomeController {
     @GetMapping("/")
     public String home() {
-        return "home";
+        return "index";
     }
```
resoures/static 폴더에 index.html 파일이 있다면, 위 로직이 굳이 필요하지 않습니다!

qxent300 replied:
네 알겠습니다.
위 내용 삭제하겠습니다.

f-lab-neo commented on build/reports/tests/test/index.html:
```diff
@@ -0,0 +1,145 @@
+<!DOCTYPE html>
+<html>
+<head>
+<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
+<meta http-equiv="x-ua-compatible" content="IE=edge"/>
+<title>Test results - Test Summary</title>
+<link href="css/base-style.css" rel="stylesheet" type="text/css"/>
+<link href="css/style.css" rel="stylesheet" type="text/css"/>
+<script src="js/report.js" type="text/javascript"></script>
+</head>
+<body>
+<div id="content">
```
build 폴더에 있는 내용들은 커밋에 포함되면 안됩니다. 확인해주시고 다시 커밋해주세요~ (git reset 을 통해 다시 커밋하시면 됩니다)

qxent300 replied:
네 알겠습니다.
build 폴더 제외하고 다시 커밋하겠습니다.

f-lab-neo reviewed (changes requested):
동식님 안녕하세요 :) 
아래와 같은 사항을 다시한번 학인해주시고 리뷰 재요청 부탁드립니다.

1. gitignore 해당되는 파일들 (build폴더 내부에 있는 파일, .gradle 폴더 파일 등등..)이 커밋에 포함되어 있는데 제거 부탁 드립니다.
2. 커밋 컨벤션에 대하여 지켜달라고 말씀드렸는데요, 확인 부탁 드립니다.
```
깃 메시지 컨벤션
https://udacity.github.io/git-styleguide/
```
3. 커밋은 기능단위로 작성부탁드립니다.
ex)
```
회원가입 기능 구현
- 회원정보(아이디, 비밀번호, 이름, 별명, 생년월일, 이메일, 전화번호, 분류)를 입력받아 DB에 저장
- 비밀번호는 SHA-256 암호화 사용
```
위와 같은 형식이 아닌

```
feat : 회원정보(아이디, 비밀번호, 이름, 별명, 생년월일, 이메일, 전화번호, 분류)를 입력받아 DB에 저장
feat : 비밀번호 SHA-256 암호화
```
위와같이 하나하나 커밋해 주세요.

커밋 이후 푸시 한 뒤, pr내용을 확인해주세요, 작업한 내용이 아닌 파일이 추가되어 있다면 무엇인가 잘못되어있는 신호입니다.

확인부탁드립니다!!
git reset + rebase 기능을 통해 커밋을 수정할 수 있습니다

f-lab-neo commented on sql/ddl.sql:
```diff
@@ -0,0 +1,11 @@
+    email VARCHAR(200),
+    phone VARCHAR(20),
+    classify VARCHAR(10)
+);
```
파일 마지막에 개행 부탁드립니다 :)

qxent300 replied:
네 알겠습니다.
개행 추가하겠습니다.

f-lab-neo commented on sql/ddl.sql:
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE MEMBER (
+    rownum INT AUTO_INCREMENT PRIMARY KEY,
+    id VARCHAR(50),
```
varchar와 char의 차이가 어떻게 될까요?

f-lab-neo commented on sql/ddl.sql:
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE MEMBER (
```
인덱스는 어떻게 걸어주는것이 좋을까요?

f-lab-neo commented on sql/ddl.sql:
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE MEMBER (
+    rownum INT AUTO_INCREMENT PRIMARY KEY,
```
Primary key가 무엇일까요?

qxent300 replied:
PK는 DB 테이블에 저장되는 데이터의 고유값으로 알고있습니다.
같은 테이블 내의 다른 데이터와 비교할 경우 컬럼에 같은 값이 저장되어도 PK가 다르면 다른 데이터로 구분 됩니다.

f-lab-neo replied:
uk를 지정 하더라도, 데이터의 고윳값을 만들어낼 수 있습니다!
uk와 pk의 차이점은 무엇일까요!?

일반적인 차이점과, innodb에서의 차이점을 함께 공부해보시면 좋을 것 같습니다! (realmysql에 내용이 있습니다!)

f-lab-neo commented on src/main/java/com/project/acre/service/MemberService.java:
```diff
@@ -0,0 +1,52 @@
+        return null;
+    }
+
+    public static class SHA256 {
+        public String encrypt(String password) throws NoSuchAlgorithmException {
+            MessageDigest md = MessageDigest.getInstance("SHA-256");
+            md.update(password.getBytes());
+            byte[] digest = md.digest();
+            return Base64.getEncoder().encodeToString(digest);
+        }
+    }
```
SHA256 클래스가 내부클래스여야 하는 이유가 있을까요!?

qxent300 replied:
처음에는 service 경로에 SHA256 클래스를 별도로 만들려고 했다가  제 생각에는 회원가입시 비밀번호를 DB에 저장하기 전에 암호화 하는 경우 외에는 다른 곳에서는 사용하지 않을 것 같아서 MemberService 안에서 내부클래스로 구현하였습니다.

f-lab-neo replied:
클래스의 사용여부에 따라 내부 클래스를 선언하신것 매우 좋습니다~!

다만 일반적으로, service에서 내부객체를 만들어 사용하는 케이스는 많이 없습니다 (결합도가 높아진다는 내용인데, 지금은 굳이 알 필요는 없습니다!). 또한 sha256과 같이 범용적인 사용성이 명확한 로직은 util성으로 따로 관리하는게 일반적입니다!

f-lab-neo commented on src/main/java/com/project/acre/service/MemberService.java:
```diff
@@ -0,0 +1,52 @@
+    public MemberDto join(MemberDto memberDto) throws NoSuchAlgorithmException {
+        Member member = new Member();
+        member.setId(memberDto.getId());
+        member.setPassword(new SHA256().encrypt(memberDto.getPassword()));
```
매번 객체를 new 하게 된다면 어떤 문제가 있을까요? 
이 문제를 어떻게 해결할 수 있을까요?

f-lab-neo commented on src/main/java/com/project/acre/service/MemberService.java:
```diff
@@ -0,0 +1,52 @@
+        }
+    }
+
+    public MemberDto join(MemberDto memberDto) throws NoSuchAlgorithmException {
```
NoSuchAlgorithenException에 대한 처리는 어떻게 하실 예정이실까요?

f-lab-neo commented on src/main/resources/templates/members/join.html:
```diff
@@ -0,0 +1,77 @@
+    });
+</script>
+</body>
+</html>
```
마지막 개행 부탁드립니다~

f-lab-neo commented on src/test/java/com/project/acre/AcreApplicationTests.java:
```diff
@@ -0,0 +1,9 @@
+package com.project.acre;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
+class AcreApplicationTests {
+
+}
```
미사용 테스트 제거 부탁드립니다!

f-lab-neo commented on src/test/java/com/project/acre/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Test
+    public void joinPasswordEncrypt() throws NoSuchAlgorithmException {
+        MemberDto memberDto = new MemberDto();
+        memberDto.setPassword(new MemberService.SHA256().encrypt("test1"));
+
+        Assertions.assertThat(memberDto).extracting("password").isEqualTo(new MemberService.SHA256().encrypt("test1"));
```
이 테스트의 목적은 무엇일까요!?

f-lab-neo reviewed (changes requested):
리뷰 확인 부탁 드립니다!

qxent300 committed:
feat: 최초 commit으로 git reset 후 코드 리팩터링

1. home.html에서 index.html로 변경
2. MemberForm.java에서 MemberDto.java로 변경
3. MySQLMemberRepository.java에서 MemberRepositoryImpl.java로 변경

qxent300 committed:
feat: 회원가입시 생일(birth)과 분류(classify) 항목 추가 및 api를 사용하도록 수정

1. 회원가입 항목 추가 및 api 사용 추가
2. 1번에 따른 MemberDto와 Member에 birth, classify 추가
3. MemberController에서 처리하던 로직을 MemberService에서 처리하도록 수정
4. MemberRepositoryImpl.java에서 join 메서드의 리턴타입 void로 수정
5. build.gradle에서 불필요한 설정 제거
6. ddl.sql에서 테이블 생성 쿼리문 수정

qxent300 committed:
feat: 회원가입시 비밀번호 SHA-256 암호화 로직 추가

qxent300 committed:
feat: MemberService에 있는 SHA-256 암호화 로직 내부클래스를 별도의 util 패키지로 분리

qxent300 committed:
feat: MemberController에서 NoSuchAlgorithmException에 대한 ExceptionHandler 처리 추가

f-lab-neo commented on src/main/java/com/project/acre/service/MemberService.java:
```diff
@@ -0,0 +1,46 @@
+
+@Service
+public class MemberService {
+    private final SHA256 sha256 = SHA256.getInstance();
```
좋습니다!, 위와같이 상태없는 객체를 계속 사용하는걸 `싱글톤 패턴` 이라고 부른답니다.
싱글톤패턴의 구현 방식들을 한번 더 공부해보시면 좋을 것 같아요!

f-lab-neo commented on src/test/java/com/project/acre/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @Test
+    public void singletonTest() {
+        SHA256 sha256_1 = SHA256.getInstance();
+        SHA256 sha256_2 = SHA256.getInstance();
+
+        Assertions.assertThat(sha256_1).isSameAs(sha256_2);
+    }
```
테스트 너무 좋네요 👍

f-lab-neo commented on src/test/java/com/project/acre/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,28 @@
+        SHA256 sha256 = SHA256.getInstance();
+        memberDto.setPassword(sha256.encrypt("test1"));
+
+        Assertions.assertThat(memberDto).extracting("password").isEqualTo(sha256.encrypt("test1"));
+    }
```
위 테스트를 보면 조금 단순화해서 보면
sha256.encrypt("test1").equals(sha256.encrypt("test1"))
를 하는것과 동일합니다.
즉, 테스트의 역할을 제대로 못하고 있어요 (1 ==1 이 true인지 확인하는게 테스트의 의미가 있을까요?)

위와같은 경우는 expect 를 실제 hash 문자열로 테스트 하는게 좋습니다.
```suggestion
        Assertions.assertThat(memberDto).extracting("password").isEqualTo("1b4f0e9851971998e732078544c96b36c3d01cedf7caa332359d6f1d83567014"); 
    }
```

다른 검증된 sha256 툴에서 test1을 암호화 한 결과를 expect로 정의함으로서, 테스트 하고자 하는 로직 (sha256.encrypt) 의 동작을 확인할 수 있게 됩니다. 만일 누군가SHA256 구현을 잘못 변경했다면, 위 테스트가 실패하면서 위험을 감지할 수 있겠지요!!

f-lab-neo reviewed (approved):
수고하셨습니다!!
사용하지 않는 controller 메서드 등 모두 제거부탁드립니다!!
다음 단계로 진행해주세요!
